ported to DPDK 19.11.3

insanum opened the pull request:
This is a huge change that breaks backwards compatibility with any prior version of DPDK. I don't expect this to be merged but wanted to make sure the effort is visible if any wants to try it out. I'm currently testing with WARP17 and DPDK 19.11.1 over a Broadcom 100Gb NIC.

insanum committed:
ported to DPDK 19.11.1

davvore33 commented on src/tpg_port.c:
```diff
@@ -168,25 +168,9 @@ static void port_setup_reta_table(uint8_t port, int nr_of_queus)
-static void port_init_tx_conf(uint16_t port)
+static void port_init_tx_conf(uint16_t port __rte_unused)
 {
-#if !defined(TPG_SW_CHECKSUMMING)
-    struct rte_eth_txconf *tx_conf;
 
-    /* We need to enable UDP/TCP hw checksum because software checksum would
-     * take too many cpu cycles, SCTP checksum offload is not supported!
-     */
-    tx_conf = &port_dev_info[port].pi_dev_info.default_txconf;
-
-    if ((tx_conf->txq_flags & ETH_TXQ_FLAGS_NOXSUMUDP) ==
-        ETH_TXQ_FLAGS_NOXSUMUDP)
-        tx_conf->txq_flags &= ~ETH_TXQ_FLAGS_NOXSUMUDP;
-    if ((tx_conf->txq_flags & ETH_TXQ_FLAGS_NOXSUMTCP) ==
-        ETH_TXQ_FLAGS_NOXSUMTCP)
-        tx_conf->txq_flags &= ~ETH_TXQ_FLAGS_NOXSUMTCP;
-#else /* !defined(TPG_SW_CHECKSUMMING) */
-    RTE_SET_USED(port);
-#endif /* !defined(TPG_SW_CHECKSUMMING) */
```
Hi @insanum ,

how did you decide to strip this part? I didn't find its replacement

davvore33 replied:
Hi @insanum ,
I went through your code change briefly and it looks good, if you could take care also at the scripts needed to install warp17 and travis configuration would be nice, if you don't know them it's fine, I can commit over your branch too

insanum replied:
> Hi @insanum ,
> 
> how did you decide to strip this part? I didn't find its replacement

Ah yes... if you look in `port_setup_port()` you'll now see a new default config section for `.txmode.offloads` that contains the flags.

insanum replied:
> Hi @insanum ,
> I went through your code change briefly and it looks good, if you could take care also at the scripts needed to install warp17 and travis configuration would be nice, if you don't know them it's fine, I can commit over your branch too

Sure thing. I'll take a look. :-)

insanum commented:
@davvore33 

There is a problem I found when trying to run with multiple cores with a single port. For the Broadcom `bnxt` driver the call to `rte_eth_dev_info_get()` returns `reta_size=0` when it's called in `port_interfaces_init()` before `port_adjust_info()`.

I don't know if this is a change in later version of DPDK, or just the behavior of `bnxt`, or just an issue with WARP17... but the size of the redirection table isn't filled in until **after** the call to `rte_eth_dev_configure()` which configures the number of rings that are going to be used.

I hacked up a workaround for my testing that hardcodes `pi_adjusted_reta_size` to 64 (I know that's what `bnxt` will return) and also added a call to `port_adjust_info()` after the call to `rte_eth_dev_configure()` in `port_setup_port()`. Any thoughts on what a proper fix should look like?

davvore33 commented:
> @davvore33
> 
> There is a problem I found when trying to run with multiple cores with a single port. For the Broadcom `bnxt` driver the call to `rte_eth_dev_info_get()` returns `reta_size=0` when it's called in `port_interfaces_init()` before `port_adjust_info()`.
> 
> I don't know if this is a change in later version of DPDK, or just the behavior of `bnxt`, or just an issue with WARP17... but the size of the redirection table isn't filled in until **after** the call to `rte_eth_dev_configure()` which configures the number of rings that are going to be used.
> 
> I hacked up a workaround for my testing that hardcodes `pi_adjusted_reta_size` to 64 (I know that's what `bnxt` will return) and also added a call to `port_adjust_info()` after the call to `rte_eth_dev_configure()` in `port_setup_port()`. Any thoughts on what a proper fix should look like?

Hi, I've never tried out a Broadcom myself before, can you please check out also with dpdk18? Moreover you can check if with the [test application](https://doc.dpdk.org/dts/test_plans/distributor_test_plan.html) provided by dpdk you see the same problem

insanum committed:
- Updated to DPDK 20.05.
    - Fixed the build_dpdk.sh script to build the igb_uio.ko driver.
    - Updated Travis config to pull and build this DPDK version.
- Show current flow-ctrl/pause setting when showing link details.
- Added TCP stats for SYN/FIN/RST packets both sent and received.
- Added RTE port drop/missed counters to the rate statistics.
- Added various debug logs for visibility in TCB purging.
- When purging established TCP sessions, when a test is stopped after its
  criteria was met, a TCP RST is now sent to the server in hopes to clean
  up its TCP state machine. When the 'clients-establ' criteria is used all
  of these TCP sessions will get purged on the client side and remain on
  the server. Sending a RST helps with this (assuming no packets are dropped
  in the Rx path on the server port, it is a huge burst). This helps
  when testing a feature like connection tracking through a switch.
  Ideally, purging TCP sessions should be graceful and send FINs while
  honoring the configured 'close' rate.

davvore33 commented:
Hi @insanum, I've seen you've updated and fixed your branch, thanks for your work but we shouldn't merge a support to a non LTS version of dpdk in our `dev/common` branch, if you prefer we can open a new branch, let's say `dev/dpdk20`
I'm although waiting for @dceara opinion too
meanwhile thanks for your help

insanum commented:
Ah yes. I totally understand. Note that while I updated the title and description to call out DPDK 20.05... no actual code changes were required between 19.11 and 20.05.

Want me to modify the description back to call out 19.11.3 and update the Travis config as well?

davvore33 commented:
Hi @insanum , sorry to bother you but if you want yes please, `19.11.3` which is the LTS version at the time we write, thanks for your help again,

dceara commented on api/warp17-stats.proto:
```diff
@@ -202,6 +202,13 @@ message TcpStatistics {
 
     required uint32 ts_reserved_bit_set  = 17;
+
+    required uint32 ts_sent_syn = 18;
```
There's no mention about these new counters in the commit logs. Do you mind mentioning them?

It might also be good to add a test case to see that these are actually incremented, please see `ts_sent_ctrl_pkts` in `test_api.py`.

insanum replied:
There is a mention in the commit log of the second commit in this pull request chain.

I added a check in the unit test for the SYN sent count. No way to check for FINs since the test criteria is based on runtime so no FINs are sent. Also no way to check for RSTs either. FWIW, for my testing I'm paying close attention to the TCP SYN/FIN/RST pkts and the stats are incrementing as expected.

dceara commented on src/tpg_tcp.c:
```diff
@@ -275,6 +275,14 @@ int tcp_close_connection(tcp_control_block_t *tcb, uint32_t flags)
     if ((flags & TCG_SILENT_CLOSE) != 0) {
+        /* don't send a RST when closing after a RST was received */
+        if (!tcb->tcb_rst_rcvd) {
+            //tpg_port_statistics_t *stats =
```
Can you please remove the commented code?

insanum replied:
Done.

dceara commented on src/tpg_tcp.c:
```diff
@@ -835,6 +843,15 @@ inline bool tcp_send_ctrl_pkt_with_sseq(tcp_control_block_t *tcb, uint32_t sseq,
 
     stats = STATS_LOCAL(tpg_tcp_statistics_t, tcb->tcb_l4.l4cb_interface);
 
+    if (flags & RTE_TCP_SYN_FLAG)
```
I'm a bit worried about the performance implications of rechecking the flags for every TCP packet. It might be better to increment the stats when we actually build the header.

insanum replied:
Yes I agree. These checks should not be made for TCP control packets that are pure ACKs only. I've modified the code flow so pure ACKs call into a separate function when sending.

dceara commented on src/tpg_tcp_sm.c:
```diff
@@ -451,16 +451,34 @@ int tsm_dispatch_net_event(tcp_control_block_t *tcb, tcpEvent_t event,
                   rte_be_to_cpu_16(pcb->pcb_tcp->rx_win),
                   rte_be_to_cpu_16(pcb->pcb_tcp->tcp_urp));
+
+        if (pcb->pcb_tcp->tcp_flags & RTE_TCP_SYN_FLAG) {
```
Same for these counters. Let's increment the stats later on in the stack.

insanum replied:
Agreed. Done.

dceara commented on src/tpg_tests.c:
```diff
@@ -1357,26 +1359,66 @@ static void test_case_purge_cbs(test_case_info_t *tc_info)
-    purge_cnt += test_purge_list(tc_info, &tc_info->tci_state.tos_to_close_cbs);
-    purge_cnt += test_purge_list(tc_info, &tc_info->tci_state.tos_to_send_cbs);
-    purge_cnt += test_purge_list(tc_info, &tc_info->tci_state.tos_closed_cbs);
+    cnt = test_purge_list(tc_info, &tc_info->tci_state.tos_to_init_cbs);
```
Is this useful information?

insanum replied:
Maybe... maybe not. It was for me. :-)
I added these purge counts and associated debug logs as I was focused on TCP connection/teardown and wanted to see the various lists TCBs lived in. This investigation is what resulted in me adding the code to (at least) send the TCP RSTs when the client test stops. I left the code here assuming someone might find it beneficial in the future as it offers more visibility.

dceara reviewed (changes requested):
Thanks for the PR!

I think the DPDK version needs to be updated in the README too. Also, it would be great if you could add all the new support (e.g., new stats and flow-control cfg) to the RELEASE_NOTES.

Thanks,
Dumitru

dceara commented:
> Hi @insanum, I've seen you've updated and fixed your branch, thanks for your work but we shouldn't merge a support to a non LTS version of dpdk in our `dev/common` branch, if you prefer we can open a new branch, let's say `dev/dpdk20`
> I'm although waiting for @dceara opinion too
> meanwhile thanks for your help

@insanum I agree, I'd prefer keeping dev/common on LTS releases. 19.11 should be fine, does that work for you?

insanum committed:
- set officially supported DPDK version to LTS 19.11.3
- exposed link Rx drops and nombuf stats via Python API (PhyStatistics)
- don't check every incoming or outgoing pkt for TCP SYN/FIN/RST flags

insanum commented:
> Hi @insanum , sorry to bother you but if you want yes please, `19.11.3` which is the LTS version at the time we write, thanks for your help again,

Yeah no problem. I agree. Done.

insanum commented:
@davvore33 @dceara 
Note that I did not update the `RELEASE_NOTES` file with any details from the changes in this pull request. Since I'm not in charge of WARP17 tagging/releases, I didn't think it was right for me to add this content. Hopefully, you two as project owners will add this content if/when a release is made. :-)

dceara commented on src/tpg_port.c:
```diff
@@ -515,32 +500,37 @@ static bool port_setup_port(uint8_t port)
             },
         },
         .txmode = {
+#if !defined(TPG_SW_CHECKSUMMING)
```
@insanum Sorry for taking so long in reviewing this PR. I finally got the chance to test it out and everything works fine except for when we compile with `TPG_SW_CHECKSUMMING` and the NIC supports TX offloading. In that case we send the mbufs with TX offload configured but the NIC is not configured for TX offload because of this `#if` here.

If I remove it then everything works fine in all cases.

I can fix it up by adding a new commit to your branch if that's ok with you. Also, do you mind if I add you to the [CONTRIBUTORS](https://github.com/Juniper/warp17/blob/dev/common/CONTRIBUTORS) file using the public information on your github profile?

Thanks,
Dumitru

dceara replied:
I decided to merge this PR as is and open a follow up to fix `TPG_SW_CHECKSUMMING`.

Thanks again for taking the time to take care of this huge change!

insanum replied:
Thank you for taking to time to review my changes and following through with merging. 👍
